mkosi/__init__.py

```diff
@@ -2645,534 +1138,197 @@ def install_unified_kernel(
 
-    os.chmod(oldpath, mode)
 
-    os.link(oldpath, newpath)
-
-    if config.no_chown:
+def qcow2_output(state: MkosiState) -> None:
+    if not state.config.output_format == OutputFormat.disk:
         return
 
-    relpath = path_relative_to_cwd(newpath)
-    chown_to_running_user(relpath)
-
-
-def link_output(state: MkosiState, artifact: Optional[BinaryIO]) -> None:
-    with complete_step("Linking image file…", f"Linked {path_relative_to_cwd(state.config.output)}"):
-        if state.config.output_format in (OutputFormat.directory, OutputFormat.subvolume):
-            if not state.root.exists():
-                return
-
-            assert artifact is None
-
-            make_read_only(state, state.root, b=False)
-            os.rename(state.root, state.config.output)
-            make_read_only(state, state.config.output, b=True)
-
-        elif state.config.output_format.is_disk() or state.config.output_format in (
-            OutputFormat.plain_squashfs,
-            OutputFormat.tar,
-            OutputFormat.cpio,
-        ):
-            if artifact is None:
-                return
-
-            _link_output(state.config, artifact.name, state.config.output)
-
+    if not state.config.qcow2:
+        return
 
-def link_output_nspawn_settings(config: MkosiConfig, path: Optional[SomeIO]) -> None:
-    if path:
-        assert config.output_nspawn_settings
-        with complete_step(
-            "Linking nspawn settings file…", f"Linked {path_relative_to_cwd(config.output_nspawn_settings)}"
-        ):
-            _link_output(config, path.name, config.output_nspawn_settings)
+    with complete_step("Converting image file to qcow2…"):
+        run(["qemu-img", "convert", "-onocow=on", "-fraw", "-Oqcow2",
+             state.staging / state.config.output.name,
+             state.staging / state.config.output.name])
 
 
-def link_output_checksum(config: MkosiConfig, checksum: Optional[SomeIO]) -> None:
-    if checksum:
-        assert config.output_checksum
-        with complete_step("Linking SHA256SUMS file…", f"Linked {path_relative_to_cwd(config.output_checksum)}"):
-            _link_output(config, checksum.name, config.output_checksum)
+def copy_nspawn_settings(state: MkosiState) -> None:
+    if state.config.nspawn_settings is None:
+        return None
 
+    with complete_step("Copying nspawn settings file…"):
+        copy_file(state.config.nspawn_settings, state.staging / state.config.output_nspawn_settings.name)
 
-def link_output_root_hash_file(config: MkosiConfig, root_hash_file: Optional[SomeIO]) -> None:
-    if root_hash_file:
-        assert config.output_root_hash_file
-        suffix = roothash_suffix(config)
-        with complete_step(f"Linking {suffix} file…", f"Linked {path_relative_to_cwd(config.output_root_hash_file)}"):
-            _link_output(config, root_hash_file.name, config.output_root_hash_file)
 
+def hash_file(of: TextIO, path: Path) -> None:
+    bs = 16 * 1024 ** 2
```

Comment:
Tiny of a style nit for readability
```suggestion
    bs = 16 * 1024**2
```